roc_jni/src/main/cpp/endpoint.cpp

```diff
@@ -0,0 +1,211 @@
+    assert(protocolClass != NULL);
+
+    if (roc_endpoint_allocate(endpoint) != 0) return -1;
+
+    tempObject = get_object_field(env, endpointClass, jendpoint, "protocol", "L" PROTOCOL_CLASS ";");
+    protocol = (roc_protocol) get_enum_value(env, protocolClass, tempObject);
+    if (roc_endpoint_set_protocol(*endpoint, protocol) != 0) return -1;
+
+    jstr = (jstring) get_object_field(env, endpointClass, jendpoint, "host", "Ljava/lang/String;");
+    host = env->GetStringUTFChars(jstr, 0);
+    assert(host != NULL);
+    if (roc_endpoint_set_host(*endpoint, host) != 0) {
+        env->ReleaseStringUTFChars(jstr, host);
+        return -1;
+    }
+    env->ReleaseStringUTFChars(jstr, host);
+
+    port = get_int_field_value(env, endpointClass, jendpoint, "port", &err);
+    if (err) return err;
+    if (roc_endpoint_set_port(*endpoint, port) != 0) return -1;
+
+    jstr = (jstring) get_object_field(env, endpointClass, jendpoint, "resource", "Ljava/lang/String;");
+    if (jstr != NULL) {
+        resource = env->GetStringUTFChars(jstr, 0);
+        if (roc_endpoint_set_resource(*endpoint, resource) != 0) {
+            env->ReleaseStringUTFChars(jstr, resource);
+            return -1;
+        }
+        env->ReleaseStringUTFChars(jstr, resource);
+    }
+    return 0;
+}
+
+void endpoint_set_port(JNIEnv *env, jobject endpoint, int port) {
+    jclass      endpointClass;
+    jfieldID    attrId;
+
+    endpointClass = env->FindClass(ENDPOINT_CLASS);
+    assert(endpointClass != NULL);
+
+    attrId = env->GetFieldID(endpointClass, "port", "I");
+    assert(attrId != NULL);
+
+    env->SetIntField(endpoint, attrId, port);
+}
+
+static const char* protocolMapping(roc_protocol protocol) {
+    switch (protocol) {
+        case ROC_PROTO_RTSP:
+            return "RTSP";
+        case ROC_PROTO_RTP:
+            return "RTP";
+        case ROC_PROTO_RTP_RS8M_SOURCE:
+            return "RTP_RS8M_SOURCE";
+        case ROC_PROTO_RS8M_REPAIR:
+            return "RS8M_REPAIR";
+        case ROC_PROTO_RTP_LDPC_SOURCE:
+            return "RTP_LDPC_SOURCE";
+        case ROC_PROTO_LDPC_REPAIR:
+            return "LDPC_REPAIR";
+        case ROC_PROTO_RTCP:
+            return "RTCP";
+        default:
+            return NULL;
+    }
+};
+
+void endpoint_set_protocol(JNIEnv *env, jobject endpoint, roc_protocol protocol) {
+    jclass      endpointClass;
+    jclass      protocolClass;
+    jfieldID    attrId;
+    jfieldID    protocolField;
+    const char* protocolValue;
+    jobject     protocolObj;
+
+    endpointClass = env->FindClass(ENDPOINT_CLASS);
+    assert(endpointClass != NULL);
+
+    protocolClass = env->FindClass(PROTOCOL_CLASS);
+    assert(protocolClass != NULL);
+
+    attrId = env->GetFieldID(endpointClass, "protocol", "L" PROTOCOL_CLASS ";");
+    assert(attrId != NULL);
+
+    protocolValue = protocolMapping(protocol); // todo: maybe better to call java function to get Enum value by roc_protocol?
```

Comment:
> maybe better to call java function to get Enum value by roc_protocol?

Sounds much better, because it current implementation we add one more place which we should keep in sync with enums.